Fix grid multicell integrity after inserting/deleting rows/columns

discnl opened the pull request:
Deal with possible size changes of a multicell (both main and inside
cells) when inserting or deleting rows or columns and, in the case of
deletion only, the disappearing of a multicell's main cell.

Closes [#4238](https://trac.wxwidgets.org/ticket/4238).

--

I tried to make the refactor changes with `wxGridCellAttrData::UpdateAttrRows`/`Cols` clear by having multiple commits. All the test changes are in one commit 28bf314879 though as I didn't really see the value of splitting it up into e.g. new additions such as helpers and then new tests making use of those.

Commit 1afb129290 with the fix and actual library changes has one very long comment for a short calculation. Perhaps instead of using the final calculation the initial more basic calculations that are also mentioned in the comment should just be used (and mention the final one in a comment).

In the tests I used Catch's `GIVEN`/`WHEN`/`THEN` which for me helped with reading test output. One drawback is they can't seem to be called using `--section` . I at least split up the tests into two `TEST_CASE_METHOD`s, one for deletions and one for insertions, to help a bit with granularity.

discnl committed:
Add basic tests for wxGrid cell attribute presence and count

Do some simple sanity checks with attributes, particularly overwriting
a cell with NULL attribute (as was already checked very usefully in
the grid sample for ref count reasons), and their total count in a grid
when inserting and deleting rows and columns.

While the tests are not particularly useful for the next intended grid
change, they do contain some functions that also work for upcoming tests
so these (harmless) tests are included as well.

discnl committed:
Add a few accelerators to menu items of grid sample

Add convenience accelerators for insert and delete menu options.

discnl committed:
Add multiple insertion support to grid sample

Only single insertions of rows and columns were supported, allow N
insertions provided the row/column selection is consecutive.
Do the same for deletions: while deleting selected rows and columns
already works it's useful to test deleting multiple positions at once.

discnl committed:
Refactor common wxGridCellAttrData::UpdateAttrRows/Cols code

Move two symmetrical functions into one and clean up, preparing it for
multicell support.

discnl committed:
Add grid tests for multicell integrity after insertions/deletions

Multicells currently don't get any special treatment when inserting
or deleting rows or columns so neither a multicell's main size nor
inside cells' sizes (which are offsets to the main cell) are updated.
Most tests fail and will be fixed by the next commit.

See #4238.

discnl committed:
Fix grid multicell integrity after inserting/deleting rows/columns

Deal with possible size changes of a multicell (both main and inside
cells) when inserting or deleting rows or columns and, in the case of
deletion only, the disappearing of a multicell's main cell.

Closes #4238.

discnl committed:
Remove doc warning regarding wxGrid multicells and deletion

Since the parent commit multicells work correctly when deleting (and
inserting) rows (or columns) intersecting with them and the comment
warning about it resulting in a crash is no longer needed.

This reverts commit 4d8e8355b4 (Added a warning about multi-cells in
wxGrid::DeleteRows() docs., 2011-12-25).

See #4238.

discnl commented:
That was hopefully the last forced push for recent mistakes in commit messages for CI fixes. If it helps feel free to cancel the running CI build for this PR (also cancelling the new one for this PR could be OK if wanted, as well as #2196).

vadz commented on tests/controls/gridtest.cpp:
```diff
@@ -1635,4 +2224,197 @@ TEST_CASE("GridBlockCoords::SymDifference", "[grid]")
+    const int rows = wxMax( m_grid->GetNumberRows(), other.GetNumberRows() );
+    const int cols = wxMax( m_grid->GetNumberCols(), other.GetNumberCols() );
+
+    int row, col;
+    for ( row = 0; row < rows; ++row )
+    {
+        for ( col = 0; col < cols; ++col )
```
This is really minor, but why not
```suggestion
    for ( int row = 0; row < rows; ++row )
    {
        for ( int col = 0; col < cols; ++col )
```

discnl replied:
Thanks, remnants from before when the matcher was matching a grid against a `Multicell` only and there was another 2D loop checking for spurious attributes.

vadz commented on tests/controls/gridtest.cpp:
```diff
@@ -1449,6 +1592,452 @@ TEST_CASE_METHOD(GridTestCase, "Grid::DrawInvalidCell", "[grid][multicell]")
+
+// Compares two grids, checking for differences with attribute presence and
+// cell sizes.
+class GridMatcher : public Catch::MatcherBase<TestableGrid>
```
I wonder if this shouldn't be renamed to `GridAttrMatcher` as I was a bit surprised to see that it didn't actually compare the grid contents.

discnl replied:
Even after reading that I was first thinking you mean it doesn't compare the properties of attributes. Ah yes, there's such a thing as strings still and cells are even considered empty without one. I'll rename it for now, at least I don't plan on making use of such functionality soon (and other grid tests may already be checking values after row/column editing, I don't know ATM).

vadz commented on tests/controls/gridtest.cpp:
```diff
@@ -31,32 +31,165 @@
-// Derive a new class inheriting from wxGrid just to get access to its
+wxString CellCoordsToString(int row, int col)
+{
+    return wxString::Format("(%d, %c)", 1 + row, 'A' + col);
```
Minor, but in the grid sample we use `R%dC%d` format which I think is also quite clear, so I'd use it here too for consistency.

If not, i.e. if we want to use the most usual format, then I'd use just "A1" format, i.e. remove the parentheses and the comma and reverse the components.

discnl replied:
I plan on using the latter because a grid's default row and column labels are shown when rendering (I find those less confusing than using only digits for the coords), so currently it's not 0-based or even using numbers for columns. The `(r, c)` format came from related trac tickets I think and documentation.

vadz commented on tests/controls/gridtest.cpp:
```diff
@@ -31,32 +31,165 @@
 public:
     explicit TestableGrid(wxWindow* parent)
-        : wxGrid(parent, wxID_ANY)
+        : wxGrid(parent, wxID_ANY), m_edit(0, 0, 0)
```
Also very minor, but maybe give `EditInfo` a default ctor instead (or just add `= 0` to the existing one)?

discnl replied:
In this case I'll add `= 0`'s. This was a later change needed because `{ 0 }` requires C++-11.

vadz commented on tests/controls/gridtest.cpp:
```diff
@@ -31,32 +31,165 @@
+    EditInfo m_edit;
+
+    // Grid as string before editing, with edit info annotated.
+    wxString m_sOriGridAnnotated;
```
It's difficult to be sure, but shouldn't this be "OrigGrid", i.e. with 2 "g"s? Also, if "s" prefix means "string", I'd remove it if you're renaming this variable anyhow because we don't use Hungarian notation normally (I wouldn't mention either of these small problems on its own, but together they just justify a comment...).

discnl replied:
The name was trying to refer to the original grid (before edits) rendered as a string, with (to be) insertions/deletions mentioned in the side-lines.
I use `ori`(ginal) myself and forgot to change this after seeing wx had `orig` usages. That's still ambiguous though (e.g. in wx `orig[inal]Width` vs. `orig[in]ScaleX`) so maybe just something with `original` or `before`. `old` could also work but that seemed less precise.

I'll drop the `s` prefix. I use it (well, not for static strings...) or `str` sometimes only for strings in usually short-lived code when the same thing is represented by different types. E.g. `width` as integer and string, where I dislike alternatives such as `widthStr`, `widthAsStr`, or `s_width` even more. Though here I suppose it can already be more clear that it's a string from the postfix which is `Annotated` right now, I'd still like to know in case there's some preference in wx for such cases.

vadz commented on tests/controls/gridtest.cpp:
```diff
@@ -31,32 +31,165 @@
+
+    void InsertRows(int pos = 0, int count = 1)
+    {
+        m_grid->DoEdit( EditInfo(pos, count, /* isRow = */ true) );
```
Not sure if it's worth changing things for, but we have `wxGridDirection` enum, i.e. you could use `wxGRID_ROW` here and `wxGRID_COLUMN` below instead of bool + comment.

discnl replied:
Nice, thanks!

vadz commented on tests/controls/gridtest.cpp:
```diff
@@ -1635,4 +2224,197 @@ TEST_CASE("GridBlockCoords::SymDifference", "[grid]")
+        else
+            InsertRows(edit.pos, edit.count);
+    }
+    else if ( edit.count < 0 )
+        DeleteCols(edit.pos, -edit.count);
+    else
+        InsertCols(edit.pos, edit.count);
```
This is also very trivial, but I'd write it symmetrically to the above, i.e.
```suggestion
    else
    {
        if ( edit.count < 0 )
            DeleteCols(edit.pos, -edit.count);
        else
            InsertCols(edit.pos, edit.count);
    }
```

Of course, if you decide to replace `isRow` with `wxGridDirection`, you'd use a switch on it instead.

discnl replied:
Thanks. It was like that, will change it back for readability.

vadz commented on interface/wx/grid.h:
```diff
@@ -2511,9 +2511,6 @@ class wxGridTableBase : public wxObject
     /**
         Delete rows from the table.
 
-        Notice that currently deleting a row intersecting a multi-cell (see
```
Do you think it could be useful to describe what happens with the multicells here? Or maybe add a separate section about multi-cells explaining this and what happens when inserting too?

discnl replied:
The name was trying to refer to the original grid (before edits) rendered as a string, with (to be) insertions/deletions mentioned in the side-lines.
I use `ori`(ginal) myself and forgot to change this after seeing wx had `orig` usages. That's still ambiguous though (e.g. in wx `orig[inal]Width` vs. `orig[in]ScaleX`) so maybe just something with `original` or `before`. `old` could also work but that seemed less precise.

I'll drop the `s` prefix. I use it (well, not for static strings...) or `str` sometimes only for strings in usually short-lived code when the same thing is represented by different types. E.g. `width` as integer and string, where I dislike alternatives such as `widthStr`, `widthAsStr`, or `s_width` even more. Though here I suppose it can already be more clear that it's a string from the postfix which is `Annotated` right now, I'd still be interested to know if there's some preference in wx for such cases.

vadz replied:
I agree that "multicell" is slang and it would probably be better to use something more formal in the manual. "merged cells" sounds good to me, but the most consistent with the names actually used in the code would be "cells spanning multiple rows/columns" which is, of course, also very long. So I basically don't know what would be the best here and so would like to leave the choice to you.

discnl replied:
Merged cells came from (not only) LibreOffice. I'll also mention the other longer term along with, probably, "merged cells".

vadz commented on src/generic/grid.cpp:
```diff
@@ -806,74 +806,181 @@ wxGridCellAttr *wxGridCellAttrData::GetAttr(int row, int col) const
+                    // Multicell is within affected range:
+                    // Adjust its size.
+
+                    if ( editCount >= 0 )
+                        mainSize += editCount;
+                    else
```
I didn't leave such comments elsewhere, but we really need braces around the branches of this if statement just because the else branch is so, so long (the guideline in wx is to use braces for anything longer than one line).

discnl replied:
I wasn't aware that included counting commented lines, particularly if it would have been just one (I'll add them there as well, and not change the comment and move it in front of the `if` or after the consequent if it fits). But yes, definitely already needed here and then apparently also around the `if` (which I wasn't sure of, thanks). I prefer always using braces anyway, having been bitten once by commenting out only the consequent line.

vadz commented on src/generic/grid.cpp:
```diff
@@ -806,74 +806,181 @@ wxGridCellAttr *wxGridCellAttrData::GetAttr(int row, int col) const
+            continue;
+        }
+
+        // Handle inside cell's existance, coords, and size.
```
```suggestion
        // Handle inside cell's existence, coords, and size.
```

discnl replied:
Ugh, thanks for spotting. Note to self: looking it up (even the correct one at that as this one looked wrong) is not enough, it also has to be actually changed :).

vadz reviewed (approved):
Thanks a lot, the changes look great, especially the new tests!

I admit I haven't fully checked the logic of the changes in `wxGrid` itself, but the tests look good and if they pass, this is good enough for me.

I think this can be merged as is, I could just make a couple of cosmetic changes myself, but please let me know if you prefer to do them.

Thanks again!

discnl commented:
Thanks for your extensive review!

> I admit I haven't fully checked the logic of the changes in `wxGrid` itself, but the tests look good and if they pass, this is good enough for me.

It would have been very surprising from my POV when they do pass here but then not elsewhere, once things got compiling anyway... The refactoring commits contain no real changes (which I could have labelled more clearly as having a start at 9813432 (!) and ending with 4df7673), and otherwise there's the library changes involving multicell editing in 1afb129 (bulky but what else to do, it didn't seem logical to me to split into e.g. fixing main and inside cells separately). These basically only extend the logic in the refactored `UpdateCellAttrRowsOrCols`, earlier on for main cells and further down for inside cells. Other cells are not affected by that commit.

For me the flow makes sense, now anyway, and it mostly feels natural what's going on (when scanning from top to bottom that is, explaining what's going on conditions-wise at a random point is more tricky) and there's (almost) no redundancy in the form of duplicate statements. This is after the code started falling into place after another major attempt (with misguided concept of how multicells work) so I'm probably biased :). This also works easily with attribute lookup changes, 'so' it seems a good fit.

Perhaps a, for once, short comment summary at the top would help, mentioning a sieve (if that's a working analogy).

> I think this can be merged as is, I could just make a couple of cosmetic changes myself, but please let me know if you prefer to do them.

Only because I noticed there's some minor new code to try out I would like to make them myself, unless you already have them (elsewhere, I suppose not) then go ahead. I would also like to put them into a single commit as to not be, IMHO, distracting and if there's the 'need' for another forced push for some reason, perhaps squash with the recent CI-related fixes. So for now I'm also not getting adventurous and check the integrated committing with GH (thanks for the commit suggestions though!). These could later be squashed as well but I'm not sure about the havoc a forced push creates on e.g. commit comments.

Let me know if it's perfectly fine to have a bunch of 'unrelated' commits though, it's just not my preference otherwise (maybe I should have thought of that before enabling that friendly looking Allow Edits option). Also if I had known I couldn't start some multi-comment edit (like when starting a review) I would have replied to this comment first instead of replying to the change comments, sorry for that. It's quite distracting this way time-spanning wise also with IRL interruptions..., and with the multiple notifications I'd imagine.

Oh and be as meticulous as you want, it's appreciated. I wanted to say pedantic here but the definition apparently doesn't exactly align with mine, which was more positive. Regarding pedantic: if at least the tests possibly can have `-Wsign-compare` I could have caught some problems earlier locally. Not sure where the setting of the warning comes from with Travis CI (compiler defaults or part of a group of settings? I couldn't easily find it in wx) but locally I need to explicitly add it for configured-based builds with apple-clang.

vadz commented:
I haven't done anything myself yet, so I'll gladly let you finish, thanks!

Concerning commits: we could just squash all of them, or we could leave a few, as you prefer. Neither would be really wrong in this particular case. But you should indeed rebase on the latest master and force push to get rid of the `wxWebRequest`-related errors in the CI. Of course, Travis CI is unusable anyhow now, we really need to finish migration to GitHub Actions a.s.a.p...

I don't know where does `-Wsign-compare` come from neither, where did you see it? I know I had to explicitly turn it off in `tests/allheaders.cpp` even though it's a potentially useful warning because there were just too many of them.

discnl commented:
Thanks for the rebase hint, I wasn't aware it was a lot better with master already and would have required upcoming related PR changes. I force pushed with a brief commit message and the CI and review changes (no doc changes) in e66c092f2b.

> I don't know where does `-Wsign-compare` come from neither, where did you see it?

I saw `-Werror=sign-compare` [here](https://travis-ci.org/github/wxWidgets/wxWidgets/jobs/756167601#L2539) first.


> I know I had to explicitly turn it off in `tests/allheaders.cpp` even though it's a potentially useful warning because there were just too many of them.

Hmm, but I was seeing the, in this case, warnings after adding `-Wsign-compare` to the tests' Makefile, even though it should be disabled later on?

vadz commented:
Thanks for the updates, but I'm afraid this has broken the GUI tests somehow, at least they're failing in all AppVeyor builds, i.e. it's not just a sporadic thing (and please ignore wxWebRequest-related failures). Does the full test suite pass for you locally? When something like this happens, it's often because a previous test has left some traces, e.g. left an orphan window behind it or something like this.

Concerning `-Wsign-compare`, I think it must be on by default.

discnl commented:
Thanks, I only had been checking on the Travis site. That certainly rings a bell: there are [TestableGrids](https://github.com/wxWidgets/wxWidgets/pull/2192/commits/99f36eb48b733a000f489e400786547ee7c4940d#diff-83eeaa12fc2066bcb0d9cf0746c178058cc004c5b7180a0f3829306970f01ea7R1711) which I didn't mind not deleting (from the matcher I suppose, where they are const) after seeing they get cleaned up anyway. In the end I didn't even hide them so yes, that already is definitely wrong.
Will look into deleting them, or perhaps having (another) `TestableGrid` just in `GridTestCase` and delete only that (I didn't want to re-use the grid if not needed, but there's only one match per test case anyway).

(also I didn't forget about the docs, it's just that it's in the category "nice to have" and wx-timewise I'd rather try to work on having less bugs for a release.)

discnl commented:
Oh and regarding the tests: I had only been running the "[grid]" tests (no problem under MSW, macOS already had some failing ones which are mostly fixable), I hardly ever run the whole suite to be honest after too many times accidentally giving user input and causing havoc with existing windows (if one of the tests themselves already didn't do that).

vadz commented:
> Oh and regarding the tests: I had only been running the "[grid]" tests (no problem under MSW, macOS already had some failing ones which are mostly fixable), I hardly ever run the whole suite to be honest after too many times accidentally giving user input and causing havoc with existing windows (if one of the tests themselves already didn't do that).

Yes, I admit I do exactly the same thing usually, but before submitting the final version it's worth running the entire test suite at least once because debugging them locally is much easier than doing it on the CI servers.

Also, I usually run the full test suite inside a VM, like this I can't just switch away from it to make sure I avoid interacting with it.

discnl commented:
Should be fixed by the (separate) last commit, as I see less tests failing (it's XP). After having to move `GridAttrMatcher` up (initially only forward declaring it) I was struggling for a while with layout of the file, such as where to place implementations of 'distracting' code. So now it's a bit of a hodgepodge...

vadz commented on samples/grid/griddemo.cpp:
```diff
@@ -1235,31 +1235,91 @@ void GridFrame::SetGridLineColour( wxCommandEvent& WXUNUSED(ev) )
-    grid->InsertRows( grid->GetGridCursorRow(), 1 );
+    int pos, insertions;
+
+    const wxArrayInt selected = grid->GetSelectedRows();
```
I was going to advise the following micro-optimization:
```suggestion
    const wxArrayInt& selected = grid->GetSelectedRows();
```
to avoid making an extra copy of a potentially big array, but then I realized that this actually shows a wrong example from efficiency point of view: if you're only interested in whether only consecutive rows are selected, you should use `GetSelectedBlocks()` and check that it has a single element, i.e.
```cpp
        const wxGridBlocks selected = m_grid->GetSelectedBlocks();
        wxGridBlocks::iterator it = selected.begin();
        if ( ++it == selected.end() )
            ... consecutive ...
```
(maybe we should also add `wxGridBlocks::GetCount()` to simplify this...)

This allows to avoid splitting the selected block(s) into rows just to check if they recombine into a single block here.

The same comment applies to all the other places where `IsConsecutiveSelection()` is used too, of course.

discnl replied:
> 
> ```c++
>         const wxGridBlocks selected = m_grid->GetSelectedBlocks();
>         wxGridBlocks::iterator it = selected.begin();
>         if ( ++it == selected.end() )
>             ... consecutive ...
> ```

Interesting suggestion, thanks!


> (maybe we should also add `wxGridBlocks::GetCount()` to simplify this...)

Not sure if `GetCount()` would be useful here at least, usually you want to do something with the iterator such as in this case. I was only interested in consecutive selections to do multiple insertions or deletions at once instead of only by one per insert/delete call.

> The same comment applies to all the other places where IsConsecutiveSelection() is used too, of course.

This would make the difference between how inserting and deleting works in the sample bigger. Currently row insertions work from either the grid cursor or you need a row selection (not cell selection) for multiple insertions. Deletion always requires a row selection.

I think what deletion requires is better (and more like how a spreadsheet program seems to work), but needing just a selection as you propose would be slightly more convenient for testing.

However if deletion should work based on block selections: how to then delete the selected rows (from all blocks) without resolving to what I think would be quite some extra code? If requiring row selection maybe `GetSelectedRowBlocks` could help.

Note that the current sample (also master) for deletion only works on row or column selections but it may appear to support based on cell selection: when I select cell B1 (not merely activate) then I can't delete selected rows, however deleting selected columns does work!?

When I move to B2 (in essence: away from the top row) then deleting the column also doesn't work. 

Conversely the same is true with columns: Selecting cell A2: can't delete column, can delete row. Move a cell selection away from the left-most column and deleting won't work again.

Not sure if that's by design and just unexpected for me, or a problem with the sample or library.

vadz replied:
Sorry, I didn't want to open a can of worms here... I thought `IsConsecutiveSelection()` could be just trivially replaced with the test `GetSelectedBlocks().GetCount() == 1`, but I didn't really check it. I am still not sure in which situation does the return value of `IsConsecutiveSelection()` differ from this expression however?

The behaviour in the sample is surprising, I don't know why does it happen neither, but I think it's something that the sample rather than the library itself does, but, again, I didn't have time to actually check it, so I could be wrong.

discnl replied:
> Sorry, I didn't want to open a can of worms here...

It was still closed when passed at least. And sorry for opening another one (which hopefully can either be discarded or go to a ticket).

> I thought `IsConsecutiveSelection()` could be just trivially replaced with the test `GetSelectedBlocks().GetCount() == 1`, but I didn't really check it. I am still not sure in which situation does the return value of `IsConsecutiveSelection()` differ from this expression however?

`IsConsecutiveSelection()` gets passed an array from `GetSelectedRows()` which works only on row selections, and there's no row selection when I merely select cells. While the expression with `GetSelectedBlocks()` works with cell selections.
So now insertions can work with just a cell selection (as mentioned, convenient but maybe not expected) but the code for deletion still requires row selections. That seems a strange difference to me and I think insertion and deletion should both either work based on row/column selections or cell selections. I'll commit and push the wip changes.

Trying to then make deletion work based on `GetSelectedBlocks()` looked problematic as mentioned. I now also checked `GetSelectedRowBlocks()` but that only returns something in `wxGridSelectRows` mode. This seems to be by design, 3307000(baad!): "[...] using wxGrid in row- or column-only selection mode". After seeing `enum wxGridSelectionModes` with bit flags (a somewhat curious combination) I wonder if this behaviour is desired. After all the grid sample starts in `wxGridSelectCells` mode which "allow[s] selecting anything", including then entire rows. 'Even' with `wxGridSelectRowsOrColumns` mode, `GetSelectedRowBlocks()` also returns an empty vector. Brute-forcing the latter to work in other modes seems to do more of what I would expect: It returns blocks when there's a row selection, just like `GetSelectedRows()` returns something. However it also returns rows based on cell selections which it probably shouldn't, similar to `GetSelectedRows()`.

If that's something to actually change (regardless if it's needed for the sample) I could open a ticket. Some interesting helper function would probably be needed for checking `wxGridSelectionModes` flags. And if `GetSelectedRowBlocks()` should only return something if selecting entire rows (and not just cells, as it could easily do now) then I suppose more code is needed beyond my current knowledge of limited exploring.


> The behaviour in the sample is surprising, I don't know why does it happen neither, but I think it's something that the sample rather than the library itself does, but, again, I didn't have time to actually check it, so I could be wrong.

I'm afraid it's the library, will try to figure out at least where the difference comes from.

discnl replied:
> > The behaviour in the sample is surprising, I don't know why does it happen neither, but I think it's something that the sample rather than the library itself does, but, again, I didn't have time to actually check it, so I could be wrong.
> 
> I'm afraid it's the library, will try to figure out at least where the difference comes from.

It's clearly the sample as it uses `IsInSelection(n, 0)` to determine deletion (hence a single cell selection in column 0 qualifies, but not in another column) instead of making use of `GetSelectedRows()` in master. Sorry for the false alarm.

vadz replied:
> `IsConsecutiveSelection()` gets passed an array from `GetSelectedRows()` which works only on row selections, and there's no row selection when I merely select cells. While the expression with `GetSelectedBlocks()` works with cell selections.

I see, thanks. In principle, I'd say that this just proves that `GetSelectedBlocks()` is the right function to use because it clearly shouldn't matter how the entire row (or several consecutive rows) got selected, what matters is that they're selected. I.e. I think the sample just does something weird now.

>  I now also checked `GetSelectedRowBlocks()` but that only returns something in `wxGridSelectRows` mode. This seems to be by design

Yes, it was. I thought they wouldn't be useful in other modes, i.e. what would be the advantage compared to `GetSelectedBlocks()` itself? And I admit I still don't see it...

> 'Even' with `wxGridSelectRowsOrColumns` mode, `GetSelectedRowBlocks()` also returns an empty vector.

... except in this one where it probably can be useful, so it probably should work in it too. OTOH I find `wxGridSelectRowsOrColumns` mode pretty useless on its own, so let's just say it's not the highest priority.

discnl replied:
> > `IsConsecutiveSelection()` gets passed an array from `GetSelectedRows()` which works only on row selections, and there's no row selection when I merely select cells. While the expression with `GetSelectedBlocks()` works with cell selections.
> 
> I see, thanks. In principle, I'd say that this just proves that `GetSelectedBlocks()` is the right function to use because it clearly shouldn't matter how the entire row (or several consecutive rows) got selected, what matters is that they're selected. I.e. I think the sample just does something weird now.

This could be terminology. Talking about the `wxGridSelectCells` mode: For me a row is selected if I click on its header. If I select a single cell then its row is not selected, I suppose it is partially but not entirely. This corresponds with `GetSelectedRows()`: there is no row selection if I select only cells.

> Yes, it was. I thought they wouldn't be useful in other modes,

Maybe I'm missing what both are trying to do, but then why is it useful for `GetSelectedRows()` to return something in `wxGridSelectCells` mode? I don't know what the function does for each mode, but I expected `GetSelectedRowBlocks()` to behave the same (perhaps because of the similarity in name).

> i.e. what would be the advantage compared to `GetSelectedBlocks()` itself? And I admit I still don't see it...

What if I want to get the selected rows (as in my definition from above, so no rows are selected when selecting cells) ? Am I supposed to filter them from `GetSelectedBlocks()` by checking the column span? And additionally for (reverse) deletion, sort the rows?
`GetSelectedRowBlocks()` seemed like it could help with both, if it behaved like `GetSelectedRows()`.

vadz replied:
> This could be terminology. Talking about the `wxGridSelectCells` mode: For me a row is selected if I click on its header. If I select a single cell then its row is not selected, I suppose it is partially but not entirely.

Yes, of course, if you select a single cell then the row is not selected. The question is what happens if you select _all_ the cells in a row one by one. IMO it's more useful to consider the row as being selected, even if it hasn't been selected all at once.

> This corresponds with `GetSelectedRows()`: there is no row selection if I select only cells.

IMO it's a problem or even a bug.

> Maybe I'm missing what both are trying to do, but then why is it useful for `GetSelectedRows()` to return something in `wxGridSelectCells` mode? I don't know what the function does for each mode, but I expected `GetSelectedRowBlocks()` to behave the same (perhaps because of the similarity in name).

As usual, there is also the question of backwards compatibility. `GetSelectedRows()` is an old function and we can't (and hopefully didn't) change its behaviour. `GetSelectedRowBlocks()` is the new function which was mostly added for efficiency reasons: `GetSelectedRows()` is unusable for large grids, while `GetSelectedRowBlocks()` definitely is.

> What if I want to get the selected rows (as in my definition from above, so no rows are selected when selecting cells) ?

Sorry, I just don't think it's the right thing to do. If you don't want to allow selecting the individual cells at all, just use the row selection mode. If you do want to allow selecting the individual cells, then it doesn't make sense to distinguish between selecting the row all at once or by selecting its individual cells because they both appear the same to the user -- so why should they behave differently?

> Am I supposed to filter them from `GetSelectedBlocks()` by checking the column span? And additionally for (reverse) deletion, sort the rows?

No, of course not.

> `GetSelectedRowBlocks()` seemed like it could help with both, if it behaved like `GetSelectedRows()`.

I think that `GetSelectedRowBlocks()` behaviour is much more useful than that of `GetSelectedRows()`.

discnl replied:
> Yes, of course, if you select a single cell then the row is not selected. The question is what happens if you select _all_ the cells in a row one by one. IMO it's more useful to consider the row as being selected, even if it hasn't been selected all at once.

Definitely, it shouldn't matter how they all got selected just that they are. And it's something `GetSelectedBlocks()` can help with, while with `GetSelectedRows()` the row is only selected if there's a single selection spanning the entire row (or clicking on its header, but that's pretty much the same under the hood).

> > This corresponds with `GetSelectedRows()`: there is no row selection if I select only cells.
> 
> IMO it's a problem or even a bug.

Reading this and some later statements I see where this went wrong now and I should have been more clear: with "cells" I meant (one,) a few or many in a grid. The only "cells" I didn't mean were all of them in a row to form an entire row span. In that latter case there had better be a row selection. Sorry for the confusion, and at least we were agreeing here...

> As usual, there is also the question of backwards compatibility. `GetSelectedRows()` is an old function and we can't (and hopefully didn't) change its behaviour. `GetSelectedRowBlocks()` is the new function which was mostly added for efficiency reasons: `GetSelectedRows()` is unusable for large grids, while `GetSelectedRowBlocks()` definitely is.

I don't know why `GetSelectedRowBlocks()` wouldn't continue that convenient behaviour of returning rows in `wxGridSelectCells` mode, it would help with it being a replacement for users wanting to use something more efficient than `GetSelectedRows()`. Regardless of whether it's useful, wxGrid has been promoting row and column selections in `wxGridSelectCells` both [in](https://github.com/wxWidgets/wxWidgets/blob/414f211b8a21ade4d7850e163070386647b4e9e0/src/generic/grid.cpp#L3741-L3745) [code](https://github.com/wxWidgets/wxWidgets/blob/414f211b8a21ade4d7850e163070386647b4e9e0/src/generic/gridsel.cpp#L169-L173) and [docs](https://github.com/wxWidgets/wxWidgets/blob/414f211b8a21ade4d7850e163070386647b4e9e0/interface/wx/grid.h#L2914-L2917).
I don't find it strange that I can do `SelectRow()` in cell mode (it probably would have been if there was something like a cells-only mode), but that I then get a different row selection depending on using `GetSelectedRows()` or `GetSelectedRowBlocks()` is unexpected to me. I don't see having new functions as a good enough reason to change behaviour here, rather it seems confusing. If I don't expect rows to be selected in cell mode then I shouldn't make use of row selection functions that could return them, so I don't think there would even be a use for an additional `GetSelectedRowBlocksDespiteBeingInSelectCellsMode()` convenience function and instead the current new ones perhaps should change.

> > Am I supposed to filter them from `GetSelectedBlocks()` by checking the column span? And additionally for (reverse) deletion, sort the rows?
> 
> No, of course not.

For basing deletions on only selected rows I think it would have been. However both insertion and deletion are just based on `GetSelectedBlocks()` now. With my previous changes it was a confusing mix of behaviour based on either `GetSelectedRows()` for the new consecutive deletions or otherwise `IsInSelection()`.

vadz replied:
I had initially decided to only handle `wxGridSelectRows` mode in `GetSelectedRowBlocks()` because there was no ambiguity about what it should do in this mode and I wasn't totally sure about the other ones and, to be frank, didn't think it was worth spending time thinking about it. However now that you've already done all the thinking, I think I agree and wouldn't mind if we change it as you propose. But this is a pure enhancement, i.e. it can be done later, without affecting any existing code, so it's not exactly the highest priority neither.

IOW let's merge this one, which fixes a real problem for now, but we can improve `GetSelected{Row,Col}Blocks()` later, of course.

discnl replied:
Sure, thanks and just an enhancement indeed. I don't see how it could get in the way and it's probably expected by long-time grid users making use of 'mixed' selections mode in some form. I could look into adding more repeated lines for the allowed modes or refactoring much later.

vadz commented on samples/grid/griddemo.cpp:
```diff
@@ -1235,31 +1235,91 @@ void GridFrame::SetGridLineColour( wxCommandEvent& WXUNUSED(ev) )
+
+    for ( int n = 0; n < grid->GetNumberRows(); )
+    {
+        if ( grid->IsInSelection( n , 0 ) )
```
Just nitpicking...
```suggestion
        if ( grid->IsInSelection( n, 0 ) )
```

discnl replied:
I only moved this code around so far (it's original code which is now the fallback option for non-consecutive selections), which I still think doesn't justify any other editing.

vadz replied:
Oops, sorry, I hadn't realized this.

vadz commented on samples/grid/griddemo.cpp:
```diff
@@ -1323,16 +1383,24 @@ void GridFrame::AutoSizeTable(wxCommandEvent& WXUNUSED(event))
+
+    for ( int n = 0; n < grid->GetNumberCols(); )
+    {
+        if ( grid->IsInSelection( 0 , n ) )
```
```suggestion
        if ( grid->IsInSelection( 0, n ) )
```

vadz reviewed (commented):
Thanks for the updates! We could merge this already because [this failure](https://github.com/wxWidgets/wxWidgets/pull/2192/checks?check_run_id=1817029881#step:10:77) looks like a fluke, and AppVeyor builds seem to work now (it hasn't quite finished yet, but the first 3 builds passed), so this problem is fixed, but please let me know if you plan to simplify the sample to use `GetSelectedBlocks()`.

Also, do you want to keep the commits or squash everything into one or something in the middle? If it's the latter, could you please rebase/recombine them as you see fit and force push?

TIA!

discnl commented:
I would like to try to simplify the sample further first. Commit-wise: probably something in the middle as this is too granular as-is for my taste (multiple commits with 'no real changes' that are just, by default, bisect noise IMHO).
I did get a useful email about the PR run failure from GH CI, just somewhat odd it comes from 'me' (while only sent to the offender, me, I assume).

vadz commented:
Sorry, just to be clear: do you plan more changes here or is this the final version? Thanks!

discnl commented:
Thanks for pinging, I was definitely planning more changes. The last wip commit was mostly there just for my recent comment. Did you perhaps not get a notification for https://github.com/wxWidgets/wxWidgets/pull/2192#discussion_r570643094 ? If not, maybe because it's outdated but the conversation wasn't resolved yet so that still seems unexpected to me (provided you Watch the repo).

discnl commented:
I've made changes to the sample and some minor edits (such as using `RnCn` coords for the grid tests), squashed here and there, and rebased on master.
I notice now insertions aren't considered consecutive if I e.g. select two neighbouring cells one by one as there are two selections. The same code for deletion (with manual run-length checking) would have to be needed but I'll leave it at this for now. In both cases `GetSelectedRowBlocks()` (or anything else efficient returning selected rows in `wxGridSelectCells` mode) would have helped as it coalesces the rows and the results are sorted, both which are also useful for the deletion part where a single for-loop could have sufficed instead of [this long mess](https://github.com/wxWidgets/wxWidgets/pull/2192/commits/30de99470a8549bf56daa8b56bab5c89de90c4c7#diff-8c629c44b05626fdf231d7db97dccdfdcfa942e429e8d90982b2296f93919a98R1278-R1327) IMHO.

vadz commented:
Merged now, thanks again for finally fixing this bug!

discnl commented:
Thanks for merging. I wanted to change (new) `multicell` refererences to `merged cell` / `mergedCell` after thinking about the docs yday but I'll save them for that docs update.
